src/hotspot/share/code/compiledIC.cpp

```diff
@@ -75,593 +67,337 @@ bool CompiledICLocker::is_safe(address code) {
+  if (UseCompressedClassPointers) {
+    return CompressedKlassPointers::decode_not_null((narrowKlass)_speculated_klass);
   } else {
-    return InlineCacheBuffer::ic_destination_for((CompiledIC *)this);
+    return (Klass*)_speculated_klass;
   }
 }
 
+//-----------------------------------------------------------------------------
+// High-level access to an inline cache. Guaranteed to be MT-safe.
 
-bool CompiledIC::is_in_transition_state() const {
-  assert(CompiledICLocker::is_safe(_method), "mt unsafe call");
-  return InlineCacheBuffer::contains(_call->destination());;
+CompiledICData* CompiledIC::data() const {
+  return _data;
 }
 
+CompiledICData* data_from_reloc_iter(RelocIterator* iter) {
+  assert(iter->type() == relocInfo::virtual_call_type, "wrong reloc. info");
+
+  virtual_call_Relocation* r = iter->virtual_call_reloc();
+  NativeMovConstReg* value = nativeMovConstReg_at(r->cached_value());
+
+  return (CompiledICData*)value->data();
+}
 
-bool CompiledIC::is_icholder_call() const {
+CompiledIC::CompiledIC(RelocIterator* iter)
+  : _method(iter->code()),
+    _data(data_from_reloc_iter(iter)),
+    _call(nativeCall_at(iter->addr()))
+{
+  assert(_method != nullptr, "must pass compiled method");
+  assert(_method->contains(iter->addr()), "must be in compiled method");
   assert(CompiledICLocker::is_safe(_method), "mt unsafe call");
-  return !_is_optimized && is_icholder_entry(ic_destination());
 }
 
-// Returns native address of 'call' instruction in inline-cache. Used by
-// the InlineCacheBuffer when it needs to find the stub.
-address CompiledIC::stub_address() const {
-  assert(is_in_transition_state(), "should only be called when we are in a transition state");
-  return _call->destination();
+CompiledIC* CompiledIC_before(CompiledMethod* nm, address return_addr) {
+  address call_site = nativeCall_before(return_addr)->instruction_address();
+  RelocIterator iter(nm, call_site, call_site + 1);
+  iter.next();
+  CompiledIC* c_ic = new CompiledIC(&iter);
+  c_ic->verify();
+  return c_ic;
 }
 
-// Clears the IC stub if the compiled IC is in transition state
-void CompiledIC::clear_ic_stub() {
-  if (is_in_transition_state()) {
-    ICStub* stub = ICStub::from_destination_address(stub_address());
-    stub->clear();
-  }
+CompiledIC* CompiledIC_at(CompiledMethod* nm, address call_site) {
+  RelocIterator iter(nm, call_site, call_site + 1);
+  iter.next();
+  CompiledIC* c_ic = new CompiledIC(&iter);
+  c_ic->verify();
+  return c_ic;
 }
 
-//-----------------------------------------------------------------------------
-// High-level access to an inline cache. Guaranteed to be MT-safe.
+CompiledIC* CompiledIC_at(Relocation* call_reloc) {
+  address call_site = call_reloc->addr();
+  CompiledMethod* cm = CodeCache::find_blob(call_reloc->addr())->as_compiled_method();
+  RelocIterator iter(cm, call_site, call_site + 1);
+  iter.next();
+  CompiledIC* c_ic = new CompiledIC(&iter);
+  c_ic->verify();
+  return c_ic;
+}
```

Comment:
Purely a style thing but could rewrite the `CompiledIC* CompiledIC_X(...)` functions in terms of each other. Made them all fit very well even on my small laptop screen.
```c++

CompiledIC* CompiledIC_before(CompiledMethod* nm, address return_addr) {
  address call_site = nativeCall_before(return_addr)->instruction_address();
  return CompiledIC_at(nm, call_site);
}

CompiledIC* CompiledIC_at(CompiledMethod* nm, address call_site) {
  RelocIterator iter(nm, call_site, call_site + 1);
  iter.next();
  return CompiledIC_at(&iter);
}

CompiledIC* CompiledIC_at(Relocation* call_reloc) {
  address call_site = call_reloc->addr();
  CompiledMethod* cm = CodeCache::find_blob(call_reloc->addr())->as_compiled_method();
  return CompiledIC_at(cm, call_site);
}

CompiledIC* CompiledIC_at(RelocIterator* reloc_iter) {
  CompiledIC* c_ic = new CompiledIC(reloc_iter);
  c_ic->verify();
  return c_ic;
}
```